### PR TITLE
[breaking?] - When using offline mirror yarn.lock does not change anymore

### PR DESCRIPTION
**Pull request by arcanis**
**Summary**

Related rfc : https://github.com/yarnpkg/rfcs/pull/51 

When enabling the offline mirror, Yarn updates the lockfile by stripping the registry URL from its `resolved` fields. This feature aims to simplify this process by making such an update unneeded.
 
Related issues: https://github.com/yarnpkg/yarn/issues/393 / https://github.com/yarnpkg/yarn/issues/394
 
 
**Test plan**

All tests should pass fine. I had a rewrite a few because of the nature of this change, but the fixtures stay the same (backward compatibility is preserved).

### CODE REVIEW

**Comment by bestander** [comment]
How does it handle backwards compatibility?
I see that you did not change existing yarn.lock files, so it should be seamless I suppose?

**Comment by arcanis** [comment]
Yep. Basically, if we have an URL, the offline cache filename is derived from the basename of this URL. If we have a path, then it is the offline cache filename. That should work with both old & new lockfiles.

**Comment by bestander** [comment]
You have lint errors and 3 tests are failing.
1 of them might be flaky though

File: package.json [hunk 1]
```diff
@@ -15,6 +15,7 @@
     "death": "^1.0.0",
     "debug": "^2.2.0",
     "detect-indent": "^5.0.0",
+    "flow": "^0.2.3",
```
Comment (bestander) [inline package.json]: why do you need flow in deps?

File: src/fetchers/git-fetcher.js [hunk 2]
```diff
@@ -15,22 +16,61 @@ const fs = require('fs');
...
 
 export default class GitFetcher extends BaseFetcher {
-  _fetch(): Promise<FetchedOverride> {
-    const {protocol, pathname} = url.parse(this.reference);
-    if (protocol === null && typeof pathname === 'string') {
-      return this.fetchFromLocal(pathname);
-    } else {
-      return this.fetchFromExternal();
+  async getLocalAvailabilityStatus(): Promise<bool> {
+    const tarballLegacyMirrorPath = this.getTarballMirrorPath({withCommit: false});
```
Comment (bestander) [inline src/fetchers/git-fetcher.js]: I think this justifies a comment explaining what is legacyMirrorPath and why we support it

File: package.json (hunk 1 above)
Comment (arcanis) [inline package.json]: I tried last friday to bump Flow's version to check if my issue was fixed in the new versions, but forgot to rollback the change.

File: src/fetchers/git-fetcher.js [hunk 3]
```diff
@@ -51,53 +91,48 @@ export default class GitFetcher extends BaseFetcher {
...
-          reject(new MessageError(reporter.lang('fetchErrorCorrupt', err.message, localTarball)));
+          reject(new MessageError(this.reporter.lang('fetchErrorCorrupt', err.message, tarballPath)));
         });
     });
   }
 
   async fetchFromExternal(): Promise<FetchedOverride> {
-    const commit = this.hash;
-    invariant(commit, 'Commit hash required');
+    const hash = this.hash; // if we don't store it in a temp variable, flow will not persist its invariant state
```
Comment (bestander) [inline src/fetchers/git-fetcher.js]: This is quite a common pattern in flow, you probably don't need to add a comment to describe it.
If someone decides to inline the variable flow won't let them anyway :)

Also I think we don't add comments on the same line with code (I might be wrong though)

**Comment by arcanis** [comment]
👍

**Comment by troyastorino** [comment]
Does this mean that the [description of the lockfile in the offline-mirror blog post](https://yarnpkg.com/blog/2016/11/24/offline-mirror/#initialize-the-new-lockfile) isn't accurate anymore?

If so, should probably either update the blog post, or port the content to [the docs page that links to the blog post](https://yarnpkg.com/lang/en/docs/offline-mirror/) and update the content there

**Comment by bestander** [comment]
Good find.
Would you want to help the project and send a PR with blog post changes?

**Comment by troyastorino** [comment]
Sure, happy to give it a go! I'm somewhat new to yarn, and I stumbled on this trying to setup the offline mirror cache for a sprawling multi-package repo (haven't gotten it working yet), so I want to make sure I understand what's going on:

* The lockfile should look the same whether or not the offline mirror is used?
  - It seems like this might be related to why [github dependencies aren't getting installed from the offline mirror cache in v0.23](https://github.com/yarnpkg/yarn/issues/3154#issuecomment-294306257) — the Resolved field looks like `https://codeload.github.com/stevemao/left-pad/tar.gz/aff6d744155a70b81f09effb8185a1564f348462` but the tarball in the offline mirror cache directory is just named `aff6d744155a70b81f09effb8185a1564f348462`. 
* A decent way to debug whether offline-mirror is working would be to run `yarn --offline` after you've cleared out your yarn cache?

**Comment by Sunyang730** [comment]
Doesn't seems like the blog post is up to date. I will create a PR for that

**Comment by bestander** [comment]
Thanks!

cc me please

**Comment by griffinmichl** [comment]
Does this solve: https://github.com/yarnpkg/yarn/issues/3154?

**Comment by bestander** [comment]
Git dependencies should still work but there can be an edge case where it broke, thanks for linking

**Comment by griffinmichl** [comment]
Another potentially related ticket: #3194

**Comment by ericgundrum** [comment]
This change has broken git+https dependencies as described in #3309. There is a mismatch between the calculated hash and the expected hash.

**Comment by bestander** [comment]
Yeah, we are on it

**Comment by steveax** [comment]
For anyone else wondering if the docs for offline mirror are up-to-date, [they were updated after this was merged](https://github.com/yarnpkg/website/commit/571273f0d14fa615e4cf20e77b75dbe692afe9a3#diff-220a2b6a12c106eb10cf45c2f00e184c)

**Pull request merged**